### ISSUE TITLE
Multiplayer Compatibility across different versions

### ISSUE THREAD

**Issue by jay20162016** [Anuken/Mindustry-Suggestions#3281]
### Describe the content or mechanics you are proposing.

There should be multiplayer compat. across versions. I understand that there will be e.g. v7 packets that are nonexistent and thus will crash in v6, however obviously complete compatibility should not be expected, but joining a v7 server from BE should work.
(if joining from outdated clients cannot be included because of incompatibility reasons, joining outdated server functionality should still be included.

Right now if you try to force a join using JS `Vars.logic.reset(); Vars.net.reset(); Vars.netClient.beginConnecting();Vars.net.connect("0.baseduser.eu.org", 6666, ()=>{print("success")})
`
you get
```
[I] Connecting to server: 0.baseduser.eu.org/143.47.182.8:6666
[I] Disconnecting.
[E] java.lang.RuntimeException: java.io.EOFException
    at arc.util.io.Reads.str(Reads.java:149)
    at mindustry.io.TypeIO.readString(TypeIO.java:542)
    at mindustry.gen.InfoToastCallPacket.handled(InfoToastCallPacket.java:30)
    at mindustry.net.Net.handleClientReceived(Net.java:264)
    at mindustry.net.ArcNetProvider$1.lambda$received$2(ArcNetProvider.java:73)
    at arc.util.TaskQueue.run(TaskQueue.java:17)
    at arc.backend.sdl.SdlApplication.loop(SdlApplication.java:162)
    at arc.backend.sdl.SdlApplication.<init>(SdlApplication.java:49)
    at mindustry.desktop.DesktopLauncher.main(DesktopLauncher.java:39)
Caused by: java.io.EOFException
    at java.base/java.io.DataInputStream.readUnsignedShort(DataInputStream.java:346)
    at java.base/java.io.DataInputStream.readUTF(DataInputStream.java:595)
    at java.base/java.io.DataInputStream.readUTF(DataInputStream.java:570)
    at arc.util.io.Reads.str(Reads.java:147)
    ... 8 more
```
 on the client
and `[I] Kicking connection www.xxx.yyy.zzz / AAAAAAAA; Reason: typeMismatch` on the server side
which seems to be an artificial barrier to joining from different versions.

### Describe how you think this content will improve the game. If you're proposing new content, mention how it may add more gameplay options or how it will fill a new niche.

This will allow players from different versions to join other versions, and will make switching versions less of a hassle.

### Before making this issue, check the boxes below to confirm that you have acknowledged them.

- [X] I have checked the [Trello](https://trello.com/b/aE2tcUwF/mindustry-trello) to make sure my suggestion isn't planned or implemented in a development version.
- [X] I am familiar with all the content already in the game or have glanced at the wiki to make sure my suggestion doesn't exist in the game yet.
- [X] I have read [README.md](https://github.com/Anuken/Mindustry-Suggestions/blob/master/README.md) to make sure my idea is not listed under the "A few things you shouldn't suggest" category.

**Comment by itcannotbe** [comment]
Why won't an artificial barrier exist? Changes in version also includes buffs/nerfs and some additions in items, furthermore game mechanic changes (see unloaders).

I wonder why it won't let you allow to connect at all.

**Comment by jay20162016** [comment]
An artificial barrier does exist. Yes, changes in items etc. are possible but you should still be able to connect to outdated servers. 

It is annoying to have 3 versions of Mindustry if you want to play on all the servers.

It doesn't allow me to connect because the versions are different.
Typemismatch corresponds to:
`server.kicked.typeMismatch = This server is not compatible with your build type.` in https://github.com/Anuken/Mindustry/blob/master/core/assets/bundles/bundle.properties#L194

**Comment by itcannotbe** [comment]
> but you should still be able to connect to outdated servers.

And cause bugs that wouldn't exist in the first place if you didn't allow them to, like the current setup

Also, if you're willing to test what will happen, compile one yourself.

**Comment by jay20162016** [comment]
> And cause bugs that wouldn't exist in the first place if you didn't allow them to, like the current setup

Of course, this would be experimental, meaning the user should expect bugs.

> Also, if you're willing to test what will happen, compile one yourself.

I did. The server kicks me because `This server is not compatible with your build type`

**Comment by itcannotbe** [comment]
> > And cause bugs that wouldn't exist in the first place if you didn't allow them to, like the current setup
> 
> Of course, this would be experimental, meaning the user should expect bugs.
> 
> > Also, if you're willing to test what will happen, compile one yourself.
> 
> I did using javascript. The server kicks me because `This server is not compatible with your build type`

Compile one yourself.

**Comment by jay20162016** [comment]
Same result as js:
`[E] java.lang.RuntimeException: java.io.EOFException
        at arc.util.io.Reads.str(Reads.java:149)
        at mindustry.io.TypeIO.readString(TypeIO.java:542)
        at mindustry.gen.InfoToastCallPacket.handled(InfoToastCallPacket.java:30)
        at mindustry.net.Net.handleClientReceived(Net.java:264)
        at mindustry.net.ArcNetProvider$1.lambda$received$2(ArcNetProvider.java:73)
        at arc.util.TaskQueue.run(TaskQueue.java:17)
        at arc.backend.sdl.SdlApplication.loop(SdlApplication.java:162)
        at arc.backend.sdl.SdlApplication.<init>(SdlApplication.java:49)
        at mindustry.desktop.DesktopLauncher.main(DesktopLauncher.java:39)
Caused by: java.io.EOFException
        at java.base/java.io.DataInputStream.readUnsignedShort(DataInputStream.java:346)
        at java.base/java.io.DataInputStream.readUTF(DataInputStream.java:595)
        at java.base/java.io.DataInputStream.readUTF(DataInputStream.java:570)
        at arc.util.io.Reads.str(Reads.java:147)
        ... 8 more
`

**Comment by jay20162016** [comment]
btw InfoToastCallPacket is 
```
   import arc.util.io.Writes;
   import java.lang.Override;
   import java.lang.String;
   import mindustry.net.Packet;

-  public class InfoToastCallPacket extends Packet {
|    private byte[] DATA = NODATA;
|
|    public String message;
|
|    public float duration;
|
|    @Override
|-   public void write(Writes WRITE) {
||     mindustry.io.TypeIO.writeString(WRITE, message);
||     WRITE.f(duration);
||   }
|
|    @Override
|-   public void read(Reads READ, int LENGTH) {
||     DATA = READ.b(LENGTH);
||   }
|
|    @Override
|-   public void handled() {
||     BAIS.setBytes(DATA);
||     message = mindustry.io.TypeIO.readString(READ);
||     duration = READ.f();
||   }
|
|    @Override
|-   public void handleClient() {
||     mindustry.ui.Menus.infoToast(message, duration);
||   }
|  }
```

**Comment by jay20162016** [comment]
RN I feel like the kick message is a server side artificial barrier.

**Comment by itcannotbe** [comment]
You know you can just compile the source and just add a version number right?

**Comment by jay20162016** [comment]
I did.

**Comment by jay20162016** [comment]
culprit is 
https://github.com/Anuken/Mindustry/blob/ec00f59b824bcccd93ad7ee3f83171ecd9eb5259/core/src/mindustry/core/NetServer.java#L207
server side

**Comment by jay20162016** [comment]
also line 262

**Comment by jay20162016** [comment]
https://github.com/Anuken/Mindustry/blob/ec00f59b824bcccd93ad7ee3f83171ecd9eb5259/core/src/mindustry/core/NetServer.java#L207
is preventing buildtype mismatches (e.g. official and bleedingedge) and custom builds if not allowed by config

 https://github.com/Anuken/Mindustry/blob/ec00f59b824bcccd93ad7ee3f83171ecd9eb5259/core/src/mindustry/core/NetServer.java#L241
is preventing different client/server versions

Basically servers can disable first check with allowcustomclients
and you can't disable second check without a custom build (all custom builds will disable second check.

**Comment by jay20162016** [comment]
well actually Foo's just sets client packet versions to = server version.

**Comment by makarhuyaker** [comment]
i keep getting error
[01-24-2022 00:55:25] [E] java.lang.RuntimeException: java.io.EOFException
         at arc.util.io.Reads.i(Reads.java:46)
         at mindustry.io.TypeIO.readBuilding(TypeIO.java:225)
         at mindustry.gen.PickedBuildPayloadCallPacket.handled(PickedBuildPayloadCallPacket.java:33)
         at mindustry.net.Net.handleServerReceived(Net.java:304)
         at mindustry.net.ArcNetProvider$2.lambda$received$2(ArcNetProvider.java:128)
         at arc.util.TaskQueue.run(TaskQueue.java:17)
         at arc.backend.headless.HeadlessApplication.mainLoop(HeadlessApplication.java:82)
         at arc.backend.headless.HeadlessApplication$1.run(HeadlessApplication.java:53)
Caused by: java.io.EOFException
         at java.io.DataInputStream.readInt(DataInputStream.java:392)
         at arc.util.io.Reads.i(Reads.java:44)
         ... 7 more

apparently not from different versions
